### PR TITLE
Fix INF bug of softmax_cross_entropy_op

### PR DESCRIPTION
**Pull request by wangxicoding**
1. Fix INF bug in softmax_cross_entropy_op.
3. Repair softmax test bug.
2. Add boundary tests to softmax_cross_entropy_op, make sure no INF.

https://github.com/PaddlePaddle/Paddle/blob/fff270eacd83384ad9a62db1af38d7db65dd2cd3/paddle/fluid/operators/softmax_with_cross_entropy_op.cu#L155
计算max_data<-64时会将其强制截断为-64。实测expf(-104.0f)时就会出现数值下溢为0, 假如softmax前面全连接层获得的logits数值都<= (-64-104)，就会出现expf数值下溢，进而导致sum(expf(logits - max_data))=0, log(0)出现-inf。
最终 sofmax -= log(0)，出现inf，导致后面反向出现inf，最终训练出现nan。

按代码意思其实应该是对(logits - max_data)做clip，避免计算loss时出现log(exp(softmax))出现log(0)的现象。数值稳定版本的softmax_cross_entropy_op见新加注释可以不需要clip。test的softmax函数已按(logits - max_data)做clip进行修改。

### CODE REVIEW

**Comment by wangxicoding** [comment]
``` python
#-*- coding:utf-8 -*-

import os
import sys
import time
import numpy as np
import paddle
import paddle.fluid as fluid

global_step = 0.0
def generator():
    for _ in range(300):
        curr_train_x = np.full((10, 4), 0 - _).astype("float32")
        res = []
        for i in range(10):
            y = i % 4
            res.append([y])
        y_true = np.array(res).astype('int64')
        yield [curr_train_x, y_true]

x = fluid.layers.data(name="x",shape=[4],dtype='float32')
y = fluid.layers.data(name="y",shape=[1],dtype='int64')

cost, prob = fluid.layers.softmax_with_cross_entropy(x, y, return_softmax=True)
avg_cost = fluid.layers.mean(cost)
sgd_optimizer = fluid.optimizer.SGD(learning_rate=0.05)
sgd_optimizer.minimize(avg_cost)

#place = fluid.CPUPlace()
place = fluid.CUDAPlace(0)
exe = fluid.Executor(place)
exe.run(fluid.default_startup_program())
for train_data, y_true in generator():
    s = time.time()
    outs = exe.run(
        feed={'x':train_data,'y':y_true},
        fetch_list=[avg_cost.name, prob])
    global_step += 1
    e = time.time()
    if global_step % 1 == 0:
        print ('iter={:.0f}, prob={}, time={}'.format(global_step, outs[1][0], e-s))
```
如这段代码，在修复前169个step时出现inf的情况，和理论上的(-64-104)出现inf吻合。
image

修复后，这段代码运行结果，正常，未出现inf
image

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 1]
```diff
@@ -150,10 +150,7 @@ static __global__ void RowReductionForMax(const T* logits_data, T* max_data,
 
   cur_max = BlockReduce<T, BlockDim>(temp_storage).Reduce(cur_max, cub::Max());
 
-  if (threadIdx.x == 0) {
-    max_data[blockIdx.x] =
-        cur_max < static_cast<T>(-64) ? static_cast<T>(-64) : cur_max;
-  }
+  if (threadIdx.x == 0) max_data[blockIdx.x] = cur_max;
```
Comment (gongweibao) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu:153]: 有没有单测能测试到这个改动的前后边界？

Comment (gavin1332) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu:153]: Have you search for any reference like caffe?

Comment (wangxicoding) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu:153]: 这个改动只涉及数值稳定版。按照数值稳定版的公式除非输入有问题，否则是不会出现INF、NAN的，
https://www.paddlepaddle.org.cn/documentation/docs/zh/1.5/api_cn/layers_cn/nn_cn.html#softmax-with-cross-entropy
image

Comment (wangxicoding) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu:153]: tensorflow按照稳定版公式写的。pytorch大概有两重保险，求max_val初值设为FLT_MAX，求log再加个保险
``` c++
    float max_val = -FLT_MAX;
    for (...) max_val = fmaxf(max_val, Xdata[idx]);  // 求max_val

    -logf(fmaxf(Pdata[idx], FLT_MIN))
```

Comment (gavin1332) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu:153]: good job

**Review by gavin1332 — DISMISSED** [review]
LGTM

**Review by gongweibao — CHANGES_REQUESTED** [review]
Need a UT

**Comment by wangxicoding** [comment]
> Need a UT

Done

**Review by gongweibao — APPROVED** [review]
LGTM
